### PR TITLE
feature1-search Shows

### PR DESCRIPTION
**Pull request by venky-kumar**
Search shows by movie language,movie year,movie genre,movie membership

### CODE REVIEW

File: sql/categoriesTable.sql [hunk 1]
```diff
@@ -0,0 +1,24 @@
+create table categories (
+	id  serial primary key,
+	genre varchar (50) not null,
+	name varchar (50) not null,
+	year int not null,
+	language varchar (50) not null,
+	category varchar(50) not null,
+	membership varchar(50) not null,
+	grade varchar(2) not null
```
Comment (venky-kumar) [inline sql/categoriesTable.sql]: check (grade in ('u','v','a'))

File: sql/categoriesTable.sql [hunk 2]
```diff
@@ -0,0 +1,24 @@
+create table categories (
+	id  serial primary key,
+	genre varchar (50) not null,
+	name varchar (50) not null,
+	year int not null,
```
Comment (venky-kumar) [inline sql/categoriesTable.sql:5]: check (year >1980)

File: src/main/java/in/venkat/dao/ShowListDao.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	 * @throws SQLException
+	 */
+	public static List<Show> getShowDetails() throws Exception, SQLException {
+		List<Show> movieList = new ArrayList<>();
+		Connection connection = null;
+		PreparedStatement preparedSt = null;
+		try {
+			connection = ConnectionUtil.getConnection();
+
+			String sql = "select * from categories ";
```
Comment (venky-kumar) [inline src/main/java/in/venkat/dao/ShowListDao.java]: select id ,genre

File: src/main/java/in/venkat/dao/ShowListDao.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+			Statement st = connection.createStatement();
```
Comment (venky-kumar) [inline src/main/java/in/venkat/dao/ShowListDao.java]: use prepared statement

File: src/main/java/in/venkat/service/ShowService.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	/**
+	 * This method is used to search movies by giving genre and language details
+	 * 
+	 * @param flimGenre
+	 * @param flimLanguage
+	 * @return
+	 * @throws SQLException
+	 * @throws Exception
+	 */
+	public static boolean searchContents(String flimGenre, String flimLanguage) throws SQLException, Exception {
```
Comment (venky-kumar) [inline src/main/java/in/venkat/service/ShowService.java]: naming

File: src/main/java/in/venkat/service/ShowService.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+		boolean isvalid = ValidateSearchDetails.validateDetails(flimGenre, flimLanguage);
```
Comment (venky-kumar) [inline src/main/java/in/venkat/service/ShowService.java]: caps

File: src/main/java/in/venkat/util/ConnectionUtil.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
+package in.venkat.util;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+
+public class ConnectionUtil {
+	private static final String DRIVER_CLASS_NAME = "org.postgresql.Driver";
```
Comment (venky-kumar) [inline src/main/java/in/venkat/util/ConnectionUtil.java]: use environmental variable

File: sql/categoriesTable.sql [hunk 8]
```diff
@@ -0,0 +1,27 @@
+create table shows (
+	id  serial primary key,
+	genre varchar (50) not null,
+	name varchar (50) not null,
+	year int not null,
+	check (year>1950),
+	language varchar (50) not null,
+	category varchar(50) not null,
+	membership varchar(50) not null,
```
Comment (csys-fresher-batch-2021) [inline sql/categoriesTable.sql]: membership - varchar(20)

File: sql/categoriesTable.sql [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+INSERT INTO categories (genre,name,year,language,category,membership,grade)
+values ('drama','bigil','2019','tamil','movie','non prime','U');
+INSERT INTO categories (id,genre,name,year,language,category,membership,grade)
+values ('drama','sivaji','2012','tamil','movie','prime','U');
+INSERT INTO categories (id,genre,name,year,language,category,membership,grade)
+values ('kids','dora','2020','english','movie','prime','v');
+INSERT INTO categories (id,genre,name,year,language,category,membership,grade)
+values ('action','explosion','2018','english','movie','prime','U');
+INSERT INTO categories (id,genre,name,year,language,category,membership,grade)
+values ('action','fast and furious','2009','english','movie','prime','U');
```
Comment (csys-fresher-batch-2021) [inline sql/categoriesTable.sql:23]: Caps "U". Table check 'u'

File: src/main/java/in/venkat/model/Show.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
+package in.venkat.model;
+
+public class Show {
+
+	private String id;
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/model/Show.java]: Use appropriate datatype for id.

File: src/main/java/in/venkat/service/ShowService.java [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+import in.venkat.validator.ValidateSearchDetails;
+
+public class ShowService {
+	private ShowService() {
+		/**
+		 * Adding private constructor
+		 */
+	}
+
+	private static final String MOVIE_ID = " MOVIE ID :";
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/service/ShowService.java:18]: Remove Space inside double quotes.

File: src/main/java/in/venkat/service/ShowService.java [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+	/**
+	 * This method is used to search movies by giving genre and language details
+	 * 
+	 * @param filmGenre
+	 * @param filmLanguage
+	 * @return
+	 * @throws SQLException
+	 * @throws Exception
+	 */
+	public static boolean searchContents(String filmGenre, String filmLanguage) throws Exception {
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/service/ShowService.java]: duplicate functionality. Can you recheck.

File: src/main/java/in/venkat/service/ShowService.java [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+	 * @throws SQLException
+	 * @throws Exception
+	 */
+	public static boolean searchMoviesByMembership(String membership) throws ClassNotFoundException, SQLException  {
+		List<Show> primeMovieList = ShowListDao.getShowDetails();
+		boolean searchDone = false;
+		for (Show show : primeMovieList) {
+			if (show.getMembership().equalsIgnoreCase(membership.trim())) {
+				Logger.log(MEMBERSHIP + membership);
+				Logger.log(MOVIE_ID + show.getId() + MOVIE_GENRE + show.getMovieGenre() + MOVIE_NAME
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/service/ShowService.java]: use toString method

File: src/main/java/in/venkat/service/ShowService.java [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+	 * @throws SQLException
+	 * @throws Exception
+	 */
+	public static boolean searchMoviesByLanguage(String filmGenre, String filmLanguage) throws Exception {
+		List<Show> primeMovieList = ShowListDao.getShowDetails();
+		boolean searchDone = false;
+		for (Show show : primeMovieList) {
+			if (show.getMovieGenre().equalsIgnoreCase(filmGenre.trim())
+					&& show.getMovieLanguage().equalsIgnoreCase(filmLanguage.trim())) {
+				Logger.log(MOVIE_GENRE + filmGenre + MOVIE_LANGUAGE + filmLanguage);
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/service/ShowService.java]: Logger.log(show)

File: src/main/java/in/venkat/util/ConnectionUtil.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+	/**
+	 * This method creates a database connection.
+	 * 
+	 * @return
+	 * @throws ClassNotFoundException
+	 * @throws SQLException
+	 */
+	public static Connection getConnection() throws SQLException, ClassNotFoundException {
+
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/util/ConnectionUtil.java:29]: add try catch ClassNotFoundException

File: src/main/java/in/venkat/validator/ValidateSearchDetails.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+		if (NameValidationUtil.validateName(filmGenre) && NameValidationUtil.validateName(filmLanguage)) {
+			valid = true;
+
+		} else {
+			throw new InvalidDetailsException("inValid details");
+		}
+		return valid;
+	}
+
+	public static boolean validateDetails(String flimLanguage)
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/validator/ValidateSearchDetails.java]: spell check - film

File: src/test/java/in/venkat/searchServiceTest/ConnectionUtilTest.java [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+
+import in.venkat.util.ConnectionUtil;
+import in.venkat.util.Logger;
+
+public class ConnectionUtilTest {
+
+	public static void main(String[] args) {
+		try {
+			ConnectionUtil.getConnection();
+			Logger.log(" Connection Success");
```
Comment (csys-fresher-batch-2021) [inline src/test/java/in/venkat/searchServiceTest/ConnectionUtilTest.java]: Connection con = ConnectionUtil.getConnection();
Logger.log("Connection Status " + (con != null ));

**Review by ghost — CHANGES_REQUESTED** [review]
Added code review comments

File: sql/Shows.sql [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+INSERT INTO shows (genre,name,year,language,category,membership,grade)
+values ('drama','bigil','2019','tamil','movie','non prime','U');
+INSERT INTO shows (genre,name,year,language,category,membership,grade)
+values ('drama','sivaji','2012','tamil','movie','prime','U');
+INSERT INTO shows (genre,name,year,language,category,membership,grade)
+values ('kids','dora','2020','english','movie','prime','V');
+INSERT INTO shows (genre,name,year,language,category,membership,grade)
+values ('action','explosion','2018','english','movie','prime','U');
+INSERT INTO shows (genre,name,year,language,category,membership,grade)
+values ('action','fast and furious','2009','english','movie','prime','U');
```
Comment (csys-fresher-batch-2021) [inline sql/Shows.sql]: Year is integer value. Remove  quotes '2009'

File: src/main/java/in/venkat/dao/ShowListDao.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+				String movieGrade = rs.getString("grade");
+				movieList.add(new Show(movieId, movieGenre, movieName, movieYear, movieLanguage, movieCategory,
+						membership, movieGrade));
+				
+				
+			}
+
+		} catch (SQLException e) {
+			e.printStackTrace();
+			throw new DbException(e, "unable ro connect");
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/dao/ShowListDao.java]: Fix spelling "unable ro connect"

File: src/main/java/in/venkat/dao/ShowListDao.java [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+
+		} finally {
+			ConnectionUtil.close(preparedSt, connection);
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/dao/ShowListDao.java]: resultset is not closed.

File: src/main/java/in/venkat/service/ShowService.java [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+		boolean isValid;
+		List<Show> search = null;
+
+		try {
+			isValid = ValidateSearchDetails.validateDetails(filmGenre, filmLanguage);
+			if (isValid) {
+				search = searchMoviesByLanguageAndGenre(filmGenre, filmLanguage);
+			}
+		} catch (EmptyFieldException | InvalidNameException | InvalidDetailsException e1) {
+			Logger.log(e1.getMessage());
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/service/ShowService.java]: Why you didn't throw exception after catching.???

File: src/main/java/in/venkat/service/ShowService.java [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+			primeMovieList = ShowListDao.getShowDetails();
+			filteredMovieList.removeAll(filteredMovieList);
+			for (Show show : primeMovieList) {
+				if (show.getMembership().equalsIgnoreCase(membership.trim())) {
+					filteredMovieList.add(show);
+					Logger.log(show);
+				}
+			}
+		} catch (DbException e) {
+			Logger.log(e.getMessage());
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/service/ShowService.java]: Why you didn't throw exception ???

File: src/main/java/in/venkat/service/ShowService.java [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+	 * @throws DbException
+	 * @throws Exception
+	 */
+	public static List<Show> searchMoviesByYear(int year) {
+		List<Show> primeMovieList;
+		List<Show> filteredMovieList = new ArrayList<>();
+
+		try {
+			primeMovieList = ShowListDao.getShowDetails();
+			filteredMovieList.removeAll(filteredMovieList);
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/service/ShowService.java]: why are you removing all movies ? 
what are you trying to do "filteredMovieList.removeAll(filteredMovieList);" ???

File: src/main/java/in/venkat/service/ShowService.java [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+			for (Show show : primeMovieList) {
+				if (show.getMovieYear() == year) {
+					filteredMovieList.add(show);
+					Logger.log(show);
+
+				}
+
+			}
+		} catch (DbException e) {
+			Logger.log(e.getMessage());
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/service/ShowService.java]: Exception should not be suppressed.

File: src/main/java/in/venkat/service/ShowService.java [hunk 25]
```diff
@@ -0,0 +1,204 @@
...
+			filteredMovieList.removeAll(filteredMovieList);
+			for (Show show : primeMovieList) {
+				if (show.getMovieLanguage().equalsIgnoreCase(filmLanguage.trim())) {
+					filteredMovieList.add(show);
+					Logger.log(show);
+
+				}
+			}
+		} catch (DbException e) {
+			Logger.log(e.getMessage());
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/service/ShowService.java]: Exception should not be suppressed.

File: src/main/java/in/venkat/service/ShowService.java [hunk 26]
```diff
@@ -0,0 +1,204 @@
...
+	 * @throws DbException
+	 * @throws Exception
+	 */
+	public static List<Show> displayAllMovies(String filmCategory) {
+		List<Show> primeMovieList;
+		List<Show> filteredMovieList = new ArrayList<>();
+
+		try {
+			primeMovieList = ShowListDao.getShowDetails();
+			filteredMovieList.removeAll(filteredMovieList);
```
Comment (csys-fresher-batch-2021) [inline src/main/java/in/venkat/service/ShowService.java]: Why are you removing all movies in the filteredMovieList ???

**Review by ghost — CHANGES_REQUESTED** [review]
Added code review comments.

File: src/main/java/in/venkat/service/ShowService.java (hunk 24 above)
Comment (venky-kumar) [reply src/main/java/in/venkat/service/ShowService.java]: resolved this by logger.exception method

Comment (venky-kumar) [reply src/main/java/in/venkat/service/ShowService.java]: not  suppressed.

File: src/main/java/in/venkat/service/ShowService.java (hunk 23 above)
Comment (venky-kumar) [reply src/main/java/in/venkat/service/ShowService.java]: removed the line (it is not used)

File: src/main/java/in/venkat/service/ShowService.java (hunk 22 above)
Comment (venky-kumar) [reply src/main/java/in/venkat/service/ShowService.java]: exception throwed

Comment (venky-kumar) [reply src/main/java/in/venkat/service/ShowService.java]: exception thrown

Comment (venky-kumar) [reply src/main/java/in/venkat/service/ShowService.java]: exception threw

File: src/main/java/in/venkat/service/ShowService.java (hunk 21 above)
Comment (venky-kumar) [reply src/main/java/in/venkat/service/ShowService.java]: fixed this

File: src/main/java/in/venkat/dao/ShowListDao.java (hunk 20 above)
Comment (venky-kumar) [reply src/main/java/in/venkat/dao/ShowListDao.java]: result set closed

File: sql/Shows.sql (hunk 18 above)
Comment (venky-kumar) [reply sql/Shows.sql]: updated the query

**Pull request merged**